### PR TITLE
Memcached ccache

### PR DESCRIPTION
**Pull request by tardyp**
This is an attempt at adding memcached backend for ccache.

The goal is to optimize cache hits in a big build cluster, without having to rely on single point of nfs failure.

I still need to make real life tests beside unit tests. This PR is for early feedback.

### CODE REVIEW

**Comment by bdrewery** [comment]
Link to OP's mailing list thread with more details: http://permalink.gmane.org/gmane.comp.compilers.ccache/1116

**Comment by tardyp** [comment]
I have been using this code in prod for 2 years now. It is building android in a memcached cluster of 150 machines

**Comment by bdrewery** [comment]
It would be nice to get it committed.

**Comment by tardyp** [comment]
if @jrosdahl is interrested to take it upstream, I can rebase it.

It is pretty big change for a very specific use of ccache so I understand if you prefer not to take it.

**Comment by itensionanders** [comment]
I have done some further work on this, including support for the "hybrid mode".
That is, it will try the local cache first - before sending requests to memcached.

The memcached format (and feature) has been kept the same as above, kudos!
Only that it has been extended with big file (>1M) support and manifest support.


Uploaded the original merge for now, will rebase the rest before the release.
See 4df050fbaea14e8ebf0e37e083def14f354bfb2f (it has some test.sh issues)

File: ccache.c [hunk 1]
```diff
@@ -299,6 +311,12 @@ get_path_in_cache(const char *name, const char *suffix)
 	char *path;
 	char *result;
 
+#if HAVE_LIBMEMCACHED
+	/* in memcached mode, we create simpler keys */
+	if (conf->memcached_conf[0] != '\0' && strcmp(suffix,".manifest") !=0 )
```
Comment (vapier) [inline ccache.c]: space after the ,

File: ccache.c [hunk 2]
```diff
@@ -1486,6 +1504,353 @@ from_cache(enum fromcache_call_mode mode, bool put_object_in_manifest)
...
 
+#ifdef HAVE_LIBMEMCACHED
+
+/* run the real compiler and put the result in memcached */
+/* TBD: split this function, to factorize the common logic with to_fscache */
+static void
+to_memcached(struct args *args)
+{
+	const char *tmp_dir = temp_dir();
+	int added_bytes=0;
```
Comment (vapier) [inline ccache.c]: spaces around the =. this comes up many times in your patches.

File: ccache.c [hunk 3]
```diff
@@ -1486,6 +1504,353 @@ from_cache(enum fromcache_call_mode mode, bool put_object_in_manifest)
...
+		   cached_dep_length=0,
+		   cached_dia_length=0;
+	/* the user might be disabling cache hits */
+	if (mode != FROMCACHE_COMPILED_MODE && conf->recache) {
+		return;
+	}
+
+	/*  Quick check if the object file is there. */
+	cached_buf = memccached_get(cached_obj,
+								&cached_obj_value,
```
Comment (vapier) [inline ccache.c]: indentation seems to have gone a bit crazy here

File: ccache.c [hunk 4]
```diff
@@ -1486,6 +1504,353 @@ from_cache(enum fromcache_call_mode mode, bool put_object_in_manifest)
...
+	    && !conf->read_only) {
+		struct stat st;
+		size_t old_size = 0; /* in bytes */
+		if (stat(manifest_path, &st) == 0) {
+			old_size = file_size(&st);
+		}
+		if (manifest_put(manifest_path, cached_obj_hash, included_files)) {
+			cc_log("Added object file hash to %s", manifest_path);
+			update_mtime(manifest_path);
+			stat(manifest_path, &st);
```
Comment (vapier) [inline ccache.c]: need to check the return value

File: ccache.h [hunk 5]
```diff
@@ -114,10 +117,23 @@ void cc_bulklog(const char *format, ...) ATTR_FORMAT(printf, 1, 2);
 void cc_log_argv(const char *prefix, char **argv);
 void fatal(const char *format, ...) ATTR_FORMAT(printf, 1, 2);
 void copy_fd(int fd_in, int fd_out);
+int safe_write(int fd_out, const char * data, size_t length);
```
Comment (vapier) [inline ccache.h]: no space after the *.  this comes up a few times in your patches.

File: memccached.c [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+static memcached_st *memc;
+static char *current_cache = NULL;
+static int current_length = 0;
+
+int memccached_init(char *conf)
+{
+    memc = memcached(conf, strlen(conf));
+    if (memc == NULL) {
+        char errorbuf[1024];
+        libmemcached_check_configuration(conf, strlen(conf), errorbuf, 1024);
```
Comment (vapier) [inline memccached.c]: use sizeof(errorbuf) instead of duplicating 1024

File: test.sh [hunk 7]
```diff
@@ -55,6 +56,11 @@ test_failed() {
     cd ..
     echo TEST FAILED
     echo "Test data and log file have been left in $TESTDIR"
+    tail -n 50 $CCACHE_LOGFILE
+    if [ ! -z $CCACHE_MEMCACHED_CONF ]; then
```
Comment (vapier) [inline test.sh]: quote the path

File: test.sh [hunk 8]
```diff
@@ -261,6 +267,9 @@ base_tests() {
     checkstat 'cache miss' 4
     compare_file reference_test1.o test1.o
 
+    if [ ! -z $CCACHE_MEMCACHED_CONF ]; then
```
Comment (vapier) [inline test.sh]: same here

File: util.c [hunk 9]
```diff
@@ -387,6 +387,27 @@ copy_file(const char *src, const char *dest, int compress_level)
 	free(tmp_name);
 	return -1;
 }
+/* Write data to a fd. */
```
Comment (vapier) [inline util.c]: add a blank line above this comment

File: util.c [hunk 10]
```diff
@@ -387,6 +387,27 @@ copy_file(const char *src, const char *dest, int compress_level)
...
+	int ret;
+	do {
+		ret = write(fd_out, data + written, length - written);
+		if (ret < 0)
+			return ret;
+		written += ret;
+	} while (written < length);
+	return 0;
+}
+/* Write data to a file. */
```
Comment (vapier) [inline util.c]: add a blank line above this comment

File: util.c [hunk 11]
```diff
@@ -387,6 +387,27 @@ copy_file(const char *src, const char *dest, int compress_level)
 	free(tmp_name);
 	return -1;
 }
+/* Write data to a fd. */
+int safe_write(int fd_out, const char * data, size_t length)
+{
+	size_t written = 0;
+	int ret;
+	do {
+		ret = write(fd_out, data + written, length - written);
```
Comment (vapier) [inline util.c]: write returns ssize_t, not int

**Comment by tardyp** [comment]
abandonned in favor of https://github.com/jrosdahl/ccache/pull/58

**Pull request closed without merging**